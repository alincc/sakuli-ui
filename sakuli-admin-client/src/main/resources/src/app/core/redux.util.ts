const NameCache = [];

export const Name = <T extends string>(n: T)=> {
  if(NameCache.indexOf(n) >= 0 ) {
    throw Error(`Name '${n}' is already in use`)
  }
  NameCache.push(n);
  return n;
}

export const notNull = (v: any): boolean => v != null;

export const log = (args: string) => (..._args: any[]) => console.log(args, ..._args);
