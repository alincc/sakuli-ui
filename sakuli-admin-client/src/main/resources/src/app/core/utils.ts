export const uniq = <T>(array: T[], idFactory: (e: T) => string) => {
  return Array.from(
    array
      .reduce((m, e) => {
        const id = idFactory(e);
        if (!m.has(id)) {
          m.set(id, e);
        }
        return m;
      }, new Map())
      .values()
  );
};

export function nothrowFn<P1, R>(cb: (p1: P1) => R, defaultValue?: R): (p1: P1) => R;
export function nothrowFn<P1, P2, R>(cb: (p1: P1, p2: P2) => R,  defaultValue?: R): (p1: P1, p2: P2) => R;
export function nothrowFn<P1, P2, P3, R>(cb: (p1: P1, p2: P2, p3: P3) => R,  defaultValue?: R): (p1: P1, p2: P2, p3: P3) => R;
export function nothrowFn(cb: (...args: any[]) => any,  defaultValue: any = undefined): (...args: any[]) => any {
  return (function (...args: any[]) {
    try {
      return cb(...args);
    } catch (e) {
      return defaultValue;
    }
  })
}
