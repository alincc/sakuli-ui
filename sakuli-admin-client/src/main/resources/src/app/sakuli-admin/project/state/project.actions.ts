import {Action} from "@ngrx/store";
import {FileResponse} from "../../../sweetest-components/services/access/model/file-response.interface";
import {Tree, TreeItem} from "../../../sweetest-components/components/presentation/tree/tree.interface";
import {ProjectModel} from "../../../sweetest-components/services/access/model/project.model";
import {actionTypeFor} from "../../../sweetest-components/services/ngrx-util/action-creator-metadata";
import {Name} from "../../../core/redux.util";

export const LOAD_PATH = 'loadpath';

interface TypedAction<T> extends Action {
  payload: T;
}

export class LoadPath implements Action {
  readonly type = LOAD_PATH;
  constructor(
    public path: string
  ) {}
}

export const APPEND_CHILDREN = 'appendchildren';
export class AppendChildren implements Action {
  readonly type = APPEND_CHILDREN;
  constructor(
    public path: string,
    public children: FileResponse[]
  ) {}
}

export const TOGGLE_OPEN = 'toggleopen';
export class ToggleOpen implements Action {
  readonly type = TOGGLE_OPEN;
  constructor(
    public item: Tree<FileResponse>
  ) {}
}

export const SELECT_FILE = 'selectfile';
export class SelectFile implements Action {
  readonly type = SELECT_FILE;
  constructor(
    public file?: Tree<FileResponse>
  ) {}
}

export const OPEN = 'open';
export class Open implements Action {
  readonly type = OPEN;
  constructor(
    public file: FileResponse
  ) {}
}

export const SET_PROJECT = 'setproject';
export class SetProject implements Action {
  readonly type = SET_PROJECT;
  constructor(
    public project: ProjectModel
  ) {}
}

export const REFRESH_PROJECT = Name('[project] REFRESH_PROJECT');
export class RefreshProject implements Action {
  readonly type = REFRESH_PROJECT;
  constructor() {}
}

export type All = LoadPath | AppendChildren | ToggleOpen | SelectFile | SetProject;
