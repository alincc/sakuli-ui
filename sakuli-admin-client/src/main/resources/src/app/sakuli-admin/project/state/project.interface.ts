import {FileResponse} from "../../../sweetest-components/services/access/model/file-response.interface";
import {Tree, TreeItem} from "../../../sweetest-components/components/presentation/tree/tree.interface";
import {ProjectModel} from "../../../sweetest-components/services/access/model/project.model";
import {createFeatureSelector, createSelector} from "@ngrx/store";
import {nothrow} from "nothrow";

export interface ProjectState {
  fileTree: Tree<FileResponse>[],
  selectedFile: Tree<FileResponse> | null,
  project: ProjectModel | null
}

export const ProjectStateInit: ProjectState = {
  fileTree: [],
  selectedFile: null,
  project: null
}


export const projectState = createFeatureSelector<ProjectState>('project');

export const project = createSelector(projectState, s => nothrow(() => s.project) || null);

export const testCases = createSelector(projectState, s=> nothrow(() => s.project.testSuite.testCases) || [])


export const projectFileRoot = createSelector(
  project,
  project => nothrow(() =>`api/files/${project.path}`) || 'api/files'
);
