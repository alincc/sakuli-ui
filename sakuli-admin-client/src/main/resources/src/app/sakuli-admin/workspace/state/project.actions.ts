import {Action} from "@ngrx/store";
import {FileResponse} from "../../../sweetest-components/services/access/model/file-response.interface";
import {Tree, TreeItem} from "../../../sweetest-components/components/presentation/tree/tree.interface";
import {ProjectModel} from "../../../sweetest-components/services/access/model/project.model";
import {uniqueName} from "../../../core/redux.util";

export const LOAD_PATH = '[testSuite] loadpath';
export class LoadPath implements Action {
  readonly type = LOAD_PATH;
  constructor(
    public path: string
  ) {}
}

export const APPEND_CHILDREN = uniqueName('[testSuite] appendchildren');
export class AppendChildren implements Action {
  readonly type = APPEND_CHILDREN;
  constructor(
    public path: string,
    public children: FileResponse[]
  ) {}
}

export const TOGGLE_OPEN = uniqueName('[testSuite] toggleopen');
export class ToggleOpen implements Action {
  readonly type = TOGGLE_OPEN;
  constructor(
    public item: Tree<FileResponse>
  ) {}
}

export const SELECT_FILE = uniqueName('[testSuite] selectfile');
export class SelectFile implements Action {
  readonly type = SELECT_FILE;
  constructor(
    readonly file?: Tree<FileResponse>
  ) {}
}

export const OPEN = uniqueName('[testSuite] open');
export class Open implements Action {
  readonly type = OPEN;
  constructor(
    readonly file: FileResponse
  ) {}
}

export const SET_PROJECT = uniqueName('[testSuite] setproject');
export class SetProject implements Action {
  readonly type = SET_PROJECT;
  constructor(
    public project: ProjectModel
  ) {}
}

export const REFRESH_PROJECT = uniqueName('[testSuite] REFRESH_PROJECT');
export class RefreshProject implements Action {
  readonly type = REFRESH_PROJECT;
  constructor() {}
}

export type All = LoadPath | AppendChildren | ToggleOpen | SelectFile | SetProject | Open;
