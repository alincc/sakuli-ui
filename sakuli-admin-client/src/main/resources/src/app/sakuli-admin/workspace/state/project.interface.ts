import {FileResponse} from "../../../sweetest-components/services/access/model/file-response.interface";
import {Tree} from "../../../sweetest-components/components/presentation/tree/tree.interface";
import {createFeatureSelector, createSelector} from "@ngrx/store";

export const WorkspaceFeatureName = 'project';

export interface WorkspaceState {
  fileTree: Tree<FileResponse>[],
  selectedFile: Tree<FileResponse> | null,
  workspace: string | null,
}

export const WorkspaceStateInit: WorkspaceState = {
  fileTree: [],
  selectedFile: null,
  workspace: null
};

const workspaceState = createFeatureSelector<WorkspaceState>(WorkspaceFeatureName);
const workspace = createSelector(
  workspaceState,
  (s) => s ? s.workspace :  ''
)

export const workpaceSelectors = {
  workspaceState,
  workspace
}
