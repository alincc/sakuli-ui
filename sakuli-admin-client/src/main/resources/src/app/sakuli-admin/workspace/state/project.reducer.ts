import {WorkspaceState, WorkspaceStateInit} from './project.interface';
import * as Actions from './project.actions';
import {absPath, FileResponse} from "../../../sweetest-components/services/access/model/file-response.interface";
import {Tree} from "../../../sweetest-components/components/presentation/tree/tree.interface";

function fileToTreeItem(file: FileResponse) {
  return ({
    ...file,
    open: false,
    busy: false,
    children: []
  })
}

function filesToTreeItems(files: FileResponse[]) {
  return files.map(fileToTreeItem);
}

function recursiveTreeWalker(filter: (item: Tree<FileResponse>) => boolean,
                             walker: (item: Tree<FileResponse>) => Tree<FileResponse>,
                             tree: Tree<FileResponse>): Tree<FileResponse> {
  tree.children = tree.children.map(child => recursiveTreeWalker(filter, walker, child));
  if (filter(tree)) {
    return walker(tree);
  }
  return tree;
}

export function projectReducer(state: WorkspaceState, action: Actions.All): WorkspaceState {
  switch (action.type) {
    case Actions.OPEN: {
      const {file} = action;
      return ({...state, workspace: absPath(file)})
    }
    case Actions.SELECT_FILE: {
      const {file: selectedFile} = action;
      return ({...state, selectedFile});
    }
    case Actions.LOAD_PATH: {
      const {path} = action;
      const fileTree = state.fileTree.map(tree => recursiveTreeWalker(
        i => path === absPath(i),
        i => ({busy: true, ...i}),
        tree
      ));
      return ({...state, fileTree});
    }
    case Actions.TOGGLE_OPEN: {
      const {item} = action;
      const fileTree = state.fileTree.map(tree => recursiveTreeWalker(
        i => absPath(i) === absPath(item),
        i => ({open: !i.open, ...i}),
        tree
      ));
      return ({...state, fileTree});
    }
    case Actions.APPEND_CHILDREN: {
      const {path, children} = action;
      const fileTree = (path === '') ? filesToTreeItems(children) :
        state.fileTree.map(tree => recursiveTreeWalker(
          i => absPath(i) === path,
          i => ({...i, children: filesToTreeItems(children)}),
          tree
        ))
      return ({...state, fileTree});
    }
  }
  return state || WorkspaceStateInit;
}
