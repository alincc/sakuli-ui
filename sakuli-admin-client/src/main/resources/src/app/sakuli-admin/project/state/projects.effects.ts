import {Injectable} from '@angular/core';
import {Actions, Effect} from '@ngrx/effects';
import {ProjectService} from '../../../sweetest-components/services/access/project.service';
import {
  AppendChildren, LOAD_PATH, LoadPath, OPEN, Open, REFRESH_PROJECT, SetProject, TOGGLE_OPEN,
  ToggleOpen
} from './project.actions';
import {of} from 'rxjs/observable/of';
import {absPath} from "../../../sweetest-components/services/access/model/file-response.interface";
import {FileService} from "../../../sweetest-components/services/access/file.service";
import {ErrorMessage} from "../../../sweetest-components/components/presentation/toast/toast-state.interface";
import {LoadTestsuite} from "../../test/state/testsuite.state";

@Injectable()
export class ProjectEffects {

  @Effect() refresh$ = this.actions$.ofType(REFRESH_PROJECT)
    .mergeMap(a => this.projectService.activeProject())
    .map(p => new SetProject(p))
    .catch(ErrorMessage('Error while fetching current project'));


  @Effect() open$ = this.actions$.ofType(LOAD_PATH)
    .mergeMap((loadPath: LoadPath) => {
      return this.fileService
        .files(loadPath.path)
        .map(fr => new AppendChildren(loadPath.path, fr))
        .catch(ErrorMessage(`Error while opening path '${loadPath.path}'`));
    });

  @Effect() afterToggle = this.actions$.ofType(TOGGLE_OPEN)
    .mergeMap((toggleOpen: ToggleOpen) => {
      return of(new LoadPath(absPath(toggleOpen.item)))
    })
    .catch(ErrorMessage(`Error while opening path`));

  @Effect() open = this.actions$.ofType(OPEN)
    .map((open: Open) => new LoadTestsuite(
      open.file.directory ? absPath(open.file) : open.file.path));

  constructor(
    private projectService: ProjectService,
    private fileService: FileService,
    private actions$: Actions
  ) {}

}
