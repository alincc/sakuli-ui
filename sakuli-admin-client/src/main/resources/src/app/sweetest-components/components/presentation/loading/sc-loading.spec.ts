import {async, ComponentFixture, inject, TestBed, tick} from "@angular/core/testing";
import {ScLoadingComponent} from "./sc-loading.component";
import {Store, StoreModule} from "@ngrx/store";
import {marbles} from 'rxjs-marbles';
import {ScLoadingModule} from "./sc-loading.module";
import {NgbModule} from "@ng-bootstrap/ng-bootstrap";
import {isLoading, loading, LoadingSetBusy, LoadingSetIdle, ScLoadingState} from "./sc-loading.state";
import {ScLoadingDisplayAs} from "./sc-loading-display-as.interface";
import {ScLoadingPresentationComponent} from "./sc-loading-presentation.component";
import {notNull} from "../../../../core/redux.util";

type LoadingAppState = { scLoading: ScLoadingState };

describe('ScLoading', () => {

  describe('Component', () => {

    let fixture: ComponentFixture<ScLoadingPresentationComponent>;

    beforeEach(async () => {
      await TestBed.configureTestingModule({
        imports: [
          ScLoadingModule,
          NgbModule.forRoot(),
          StoreModule.forRoot({})
        ],

      }).compileComponents();
      fixture = TestBed.createComponent(ScLoadingPresentationComponent);
    });

    ([
      ['progressbar', true],
      ['progressbar', false],
      ['spinner', true],
      ['spinner', false],
    ] as ([ScLoadingDisplayAs, boolean])[]).forEach(([displayAs, show]) => {

      it(`should fullfill snapshot for ${displayAs} and show: ${show}`, () => {
        fixture.componentInstance.displayAs = displayAs;
        fixture.componentInstance.for = 'test';
        fixture.componentInstance.show = show;
        fixture.detectChanges();
        expect(fixture).toMatchSnapshot();
      })
    })

  });

  describe('state', () => {
    let store: Store<LoadingAppState>;

    beforeEach(async () => {
      await TestBed.configureTestingModule( {
        imports: [
          ScLoadingModule,
          StoreModule.forRoot({})
        ]
      }).compileComponents();
      store = TestBed.get(Store);
    });

    it('should set a state busy', marbles( (m) => {
      store.dispatch(new LoadingSetBusy('test'));
      const expected = m.cold('a', {a: true});
      const selection = store.select(isLoading('test')).filter(notNull);
      m.expect(selection).toBeObservable(expected);
    }))

    it('should set a state idle', marbles( (m) => {
      store.dispatch(new LoadingSetIdle('test'));
      const expected = m.cold('a', {a: false});
      const selection = store.select(isLoading('test')).filter(notNull);
      m.expect(selection).toBeObservable(expected);
    }))
  })


});
