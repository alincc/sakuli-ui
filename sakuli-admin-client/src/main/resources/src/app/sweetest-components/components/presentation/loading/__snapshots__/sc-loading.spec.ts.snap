// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScLoading Component should fullfill snapshot for progressbar and show: false 1`] = `
<sc-loading-presentation
  displayAs={[Function String]}
  for={[Function String]}
  show="false"
>
      
      
    
</sc-loading-presentation>
`;

exports[`ScLoading Component should fullfill snapshot for progressbar and show: true 1`] = `
<sc-loading-presentation
  displayAs={[Function String]}
  for={[Function String]}
  show={[Function Boolean]}
>
      
      
        
        
          <ngb-progressbar
    ng-reflect-animated="true"
    ng-reflect-striped="true"
    ng-reflect-value="100"
    value="100"
  >
    
      
    <div
      class="progress"
    >
      
        
      <div
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="100"
        class="progress-bar progress-bar-animated progress-bar-striped"
        role="progressbar"
        style="width: 100%;"
      >
        
          
        
        
            
        
          
        
        
      </div>
      
      
    </div>
    
    
  </ngb-progressbar>
        
      
    
</sc-loading-presentation>
`;

exports[`ScLoading Component should fullfill snapshot for spinner and show: false 1`] = `
<sc-loading-presentation
  displayAs={[Function String]}
  for={[Function String]}
  show="false"
>
      
      
    
</sc-loading-presentation>
`;

exports[`ScLoading Component should fullfill snapshot for spinner and show: true 1`] = `
<sc-loading-presentation
  displayAs={[Function String]}
  for={[Function String]}
  show={[Function Boolean]}
>
      
      
        
          <sc-icon
    _nghost-c2=""
    class="d-inline-block "
    icon="fa-spinner"
    ng-reflect-icon="fa-spinner"
    ng-reflect-spin="true"
  >
    
      
    <div
      _ngcontent-c2=""
      class="w-100 flex-row d-flex flex-nowrap align-content-center align-items-center"
    >
      
        
      <i
        _ngcontent-c2=""
        class="fa fa-spin fa-spinner"
        ng-reflect-klass="fa"
        ng-reflect-ng-class="[object Object]"
        ng-reflect-ng-style="[object Object]"
      />
      
        
      <div
        _ngcontent-c2=""
        class="d-flex w-100 justify-content-between align-items-center"
      >
        
          
        
            
        
          
        
        
      </div>
      
      
    </div>
    
    
  </sc-icon>
        
        
      
    
</sc-loading-presentation>
`;
